File: Data Structures & Algorithms/search-for-word/submission-2.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n, m = len(board), len(board[0])

        def rec(r, c, i):
            if i == len(word):
                return True
            if (r >= n or c >= m or r < 0 or c <0):
                return False
            if board[r][c] != word[i]:
                return False
            i += 1
            t = board[r][c]
            board[r][c] = '#'
            res = (
                rec(r+1, c, i)
                or 
                rec(r, c+1, i)
                or 
                rec(r-1, c, i)
                or
                rec(r, c-1, i)
            )
            board[r][c] = t
            return res

        for r in range(n):
            for c in range(m):
                if rec(r, c, 0):
                    return True
        return False
        

File: Data Structures & Algorithms/anagram-groups/submission-2.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<List<Integer>, List<String>> map = new HashMap<>();
        for (String s : strs) {
            List<Integer> key = wordToList(s);
            map.putIfAbsent(key, new ArrayList<>());
            map.get(key).add(s);
        }
        return new ArrayList<List<String>>(map.values());
    }

    public List<Integer> wordToList(String word) {
        List<Integer> res = new ArrayList<>(26);
        for (int i = 0; i < 26; i++) {
            res.add(0);
        }
        for (char c : word.toCharArray()) {
            res.set(c - 'a', res.get(c - 'a') + 1);
        }
        return res;
    }
}
